[WebGPU] Convert all adds and muls into checked arithmetic - part 1

mwyrzykowski opened the pull request:
#### be64ae4a8627494d9ea27c261bc09c573f1c2e7d
<pre>
[WebGPU] Convert all adds and muls into checked arithmetic - part 1
<a href="https://bugs.webkit.org/show_bug.cgi?id=281365">https://bugs.webkit.org/show_bug.cgi?id=281365</a>
<a href="https://rdar.apple.com/137793734">rdar://137793734</a>

Reviewed by Geoffrey Garen.

Convert CommandEncoder.mm, Queue.mm, and Texture.mm to
exclusively use checked arithmetic for all values passed
over IPC.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::copyBufferToTexture):
(WebGPU::CommandEncoder::clearTextureIfNeeded):
(WebGPU::CommandEncoder::copyTextureToBuffer):
(WebGPU::CommandEncoder::copyTextureToTexture):
(WebGPU::validateResolveQuerySet):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeTexture):
(WebGPU::Queue::clearTextureViewIfNeeded):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::bytesPerRow):

Canonical link: <a href="https://commits.webkit.org/285246@main">https://commits.webkit.org/285246@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db380c9ed9b39bcfa13e750bd97d8666936f4664

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71735 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51148 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24509 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75850 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22940 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58949 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22760 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56637 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15126 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74801 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46399 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61793 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37086 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43060 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21281 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64966 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19623 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77569 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15969 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18806 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64356 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16013 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61826 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64362 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15906 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12523 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6164 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46948 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48019 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49303 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47761 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -910,13 +910,15 @@ static bool refersToAllAspects(WGPUTextureFormat format, WGPUTextureAspect aspec
-        sourceBytesPerRow = std::min<uint32_t>(sourceBytesPerRow, blockSize * m_device->limits().maxTextureDimension1D);
-        break;
+    case WGPUTextureDimension_1D: {
+        auto checkedBlockSizeTimesTextureDimension = checkedProduct<uint32_t>(blockSize, m_device->limits().maxTextureDimension1D);
```
these probably technically aren't needed, since neither value originates from or is impacted by values passed over IPC

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -947,44 +949,64 @@ static bool refersToAllAspects(WGPUTextureFormat format, WGPUTextureAspect aspec
     }
 
-    NSUInteger maxSourceBytesPerRow = textureDimension == WGPUTextureDimension_3D ? (2048 * blockSize) : sourceBytesPerRow;
+    auto blockSizeTimes2048 = checkedProduct<NSUInteger>(2048, blockSize);
```
missing if (...) return

mwyrzykowski replied:
actually I will revert this, it is impossible to approach an overflow here (blockSize is from a function that returns at most 16)

mwyrzykowski commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -2157,7 +2157,8 @@ static bool textureViewFormatCompatible(WGPUTextureFormat format1, WGPUTextureFo
         return 0;
     }
     NSUInteger blocksInWidth = textureWidth / blockWidth;
-    return Texture::texelBlockSize(format) * blocksInWidth * sampleCount;
```
likely this is not possible to overflow, since `texelBlockSize` and `texelBlockWidth` can be at most `16`. There is a risk someone passes `sampleCount` without validating it, but it is also an error to pass anything other than `1` or `4` for sample count.

Seems fine to leave the checked arithmetic however

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -1444,9 +1513,23 @@ static bool hasValidDimensions(WGPUTextureDimension dimension, NSUInteger width,
+            auto layerTimesDestinatinBytesPerImage = checkedProduct<NSUInteger>(layer, destinationBytesPerImage);
+            if (layerTimesDestinatinBytesPerImage.hasOverflowed())
+                return;
+            auto checkedDestinationOffset = checkedSum<NSUInteger>(destination.layout.offset + layerTimesDestinatinBytesPerImage.value());
```
this line is wrong, contains ` + ` instead of two arguments, will fix

geoffreygaren commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -910,13 +910,15 @@ static bool refersToAllAspects(WGPUTextureFormat format, WGPUTextureAspect aspec
     auto aspectSpecificFormat = Texture::aspectSpecificFormat(destinationTexture.format(), destination.aspect);
     auto blockSize = Texture::texelBlockSize(aspectSpecificFormat);
```
Did you consider declaring blockSize to be `Checked<uint32_t>`, or even changing Texture and related types to return `Checked<T>` for every integral type it currently returns?

If the type is `Checked<T>` from the start, then the compiler will track where it flows for you.

mwyrzykowski replied:
that is a great idea, while `texelBlockSize` will never be greater than `16`, it is often used in computations involving `width`, etc so then we ensure those are checked

geoffreygaren commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -2157,10 +2157,11 @@ static bool textureViewFormatCompatible(WGPUTextureFormat format1, WGPUTextureFo
     }
     NSUInteger blocksInWidth = textureWidth / blockWidth;
-    return Texture::texelBlockSize(format) * blocksInWidth * sampleCount;
+    auto product = checkedProduct<NSUInteger>(Texture::texelBlockSize(format), blocksInWidth, sampleCount);
```
I wonder if `Texture::texelBlockWidth()` and all the data members of `m_device->limits()` should changed to return Checked<T>?

I also wonder if you want to use `Checked<T, RecordOverflow>`, so that instead of writing `checkedProduct(a, b)` you can just write `a * b`?

If typing `RecordOverflow` everywhere is annoying, you can always add more convenience typedefs to CheckedArithmetic.h, like CheckedNSUInteger. (CheckedUint32 is already there.)

geoffreygaren reviewed (approved):
r=me

This patch looks good to me as is, but do consider just going wild and using Checked<T, RecordOverflow> at the declaration site for all values that might be involved in validation. It's just a lot easier to track value types at the point of ingress that it is to track all value uses.

mwyrzykowski committed:
[WebGPU] Convert all adds and muls into checked arithmetic - part 1
https://bugs.webkit.org/show_bug.cgi?id=281365
rdar://137793734

Reviewed by Geoffrey Garen.

Convert CommandEncoder.mm, Queue.mm, and Texture.mm to
exclusively use checked arithmetic for all values passed
over IPC.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::copyBufferToTexture):
(WebGPU::CommandEncoder::clearTextureIfNeeded):
(WebGPU::CommandEncoder::copyTextureToBuffer):
(WebGPU::CommandEncoder::copyTextureToTexture):
(WebGPU::validateResolveQuerySet):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeTexture):
(WebGPU::Queue::clearTextureViewIfNeeded):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::bytesPerRow):

Canonical link: https://commits.webkit.org/285246@main
